File: src/main/java/virtualMachine/stack/vm_instruction_parsing/VMInstructionParser.java
package virtualMachine.stack.vm_instruction_parsing;

import virtualMachine.stack.datawrappers.instruction.Commands;
import virtualMachine.stack.datawrappers.instruction.Instruction;
import virtualMachine.stack.memory.GlobalVirtualMemory;
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.AllInstructionProcessors;
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.InstructionProcessor;

import java.util.List;

import static virtualMachine.stack.datawrappers.instruction.Commands.*;

public class VMInstructionParser {

    private final List<Instruction> vmInstructions;
    private final GlobalVirtualMemory globalVirtualMemory = new GlobalVirtualMemory();
    private final AllInstructionProcessors instructionProcessors = new AllInstructionProcessors();

    public VMInstructionParser(List<Instruction> vmInstructions) {
        this.vmInstructions = vmInstructions;

        initializeFunctions(vmInstructions);
    }

    public void processInstruction(Instruction instruction) {
        var instructionProcessor = getInstructionProcessorFromCommand(instruction);
        instructionProcessor.processInstruction(instruction, globalVirtualMemory);
    }

    public void executeVmInstructions() {
        while (globalVirtualMemory.hasNextInstruction(vmInstructions)) {
            int instructionPointer = globalVirtualMemory.nextInstruction();

            Instruction instruction = vmInstructions.get(instructionPointer);
            var instructionProcessor = getInstructionProcessorFromCommand(instruction);
            instructionProcessor.processInstruction(instruction, globalVirtualMemory);
        }

    }

    private InstructionProcessor getInstructionProcessorFromCommand(Instruction instruction) {
        Commands command = instruction.getCommand();

        if (command == POP) {

            return instructionProcessors.getPopInstructionProcessor();
        } else if (command == PUSH) {

            return instructionProcessors.getPushInstructionProcessor();
        } else if (Commands.allBinaryCommands().contains(command)) {

            return instructionProcessors.getBinaryInstructionProcessor();
        } else if (Commands.allUnaryCommands().contains(command)) {

            return instructionProcessors.getUnaryInstructionProcessor();
        } else if (Commands.allFunctionCommands().contains(command)) {

            return instructionProcessors.getFunctionProcessor();
        } else if (Commands.programFlowCommands().contains(command)) {

            return instructionProcessors.getProgramFlowProcessor();
        } else {
            throw new RuntimeException("Unknown instruction: " + instruction);
        }
    }

    public GlobalVirtualMemory getVirtualMemory() {
        return globalVirtualMemory;
    }


    private void initializeFunctions(List<Instruction> vmInstructions) {
        for (int i = 0; i < vmInstructions.size(); i++) {
            var instruction = vmInstructions.get(i);
            if (instruction.getCommand() == FUNCTION) {
                globalVirtualMemory.addFunctionInstructionLocation(instruction.getOperand(), i);
            } else if (instruction.getCommand() == LABEL) {
                globalVirtualMemory.addLabel(instruction.getOperand(), i);
            }
        }
    }
}

File: src/main/java/virtualMachine/stack/vm_instruction_parsing/VMInstructionParser.kt
package virtualMachine.stack.vm_instruction_parsing

import virtualMachine.stack.datawrappers.instruction.Commands
import virtualMachine.stack.datawrappers.instruction.Instruction
import virtualMachine.stack.memory.GlobalVirtualMemory
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.InstructionProcessor
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.impl.BinaryInstructionProcessor
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.impl.PopInstructionProcessor
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.impl.PushInstructionProcessor
import virtualMachine.stack.vm_instruction_parsing.vm_instruction_processing.impl.UnaryInstructionProcessor

// <command> <datatype> <value>
fun getValueFromCommand(instruction: String) : String {
    if (instruction.contains(" ")) {
        val splitCommand: List<String> = instruction.split(" ")
        return splitCommand[splitCommand.size - 1]
    }
    return instruction
}

private val allBinaryInstructions: Set<String> = setOf("add", "sub", "eq", "gt", "lt", "and", "or")
private val allUnaryInstructions: Set<String> = setOf("neg", "not")

class VMInstructionParser {

    private val virtualMemory = GlobalVirtualMemory()

//    fun processInstruction(instruction: String) {
//        val processor : InstructionProcessor = getInstructionProcessor(instruction)
//        processor.processInstruction(instruction, virtualMemory)
//    }

    fun processInstruction(instruction: Instruction) {
        val processor : InstructionProcessor = getInstructionProcessor(instruction)
        processor.processInstruction(instruction, virtualMemory)
    }

    private fun getInstructionProcessor(instruction: Instruction) : InstructionProcessor {
        return when {
            instruction.command == Commands.POP -> PopInstructionProcessor()
            instruction.command == Commands.PUSH -> PushInstructionProcessor()
            Commands.allBinaryCommands().contains(instruction.command) -> BinaryInstructionProcessor()
            Commands.allUnaryCommands().contains(instruction.command) -> UnaryInstructionProcessor()
            else -> throw RuntimeException("Unknown instruction $instruction")
        }
    }

//    private fun getInstructionProcessor(instruction: String) : InstructionProcessor {
//        return when {
//            instruction.contains("pop") -> PopInstructionProcessor()
//            instruction.contains("push") -> PushInstructionProcessor()
//            allBinaryInstructions.contains(instruction) -> BinaryInstructionProcessor()
//            allUnaryInstructions.contains(instruction) -> UnaryInstructionProcessor()
//            else -> throw RuntimeException("Unknown instruction $instruction")
//        }
//    }

    fun getVirtualMemory() : GlobalVirtualMemory {
        return virtualMemory
    }

}